# EDA(Exploratory Data Analysis) - group 11

## load packages

In [1]:
import numpy as np
import pandas as pd
import altair as alt

## load data

In [2]:
data = pd.read_csv("data/cricket_test.csv")
data = data.drop(columns = ["Unnamed: 0"])
data

game_id   season     team  over          batter batter_id       bowler  \
0      211028     2005  England     0  ME Trescothick  ea42ddb9        B Lee   
1      211028     2005  England     0  ME Trescothick  ea42ddb9        B Lee   
2      211028     2005  England     0        GO Jones  2e929b99        B Lee   
3      211028     2005  England     0        GO Jones  2e929b99        B Lee   
4      211028     2005  England     0        GO Jones  2e929b99        B Lee   
...       ...      ...      ...   ...             ...       ...          ...   
2305   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2306   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2307   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2308   255954  2006/07    India    19        SK Raina  1dc12ab9  RJ Peterson   
2309   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   

     bowler_id     non_striker non_striker_id  ...  byes  wicket  player_out  \
0     dd09ff8e        GO Jones       2e929b99  ...     0       0         NaN   
1     dd09ff8e        GO Jones       2e929b99  ...     0       0         NaN   
2     dd09ff8e  ME Trescothick       ea42ddb9  ...     0       0         NaN   
3     dd09ff8e  ME Trescothick       ea42ddb9  ...     0       0         NaN   
4     dd09ff8e  ME Trescothick       ea42ddb9  ...     0       0         NaN   
...        ...             ...            ...  ...   ...     ...         ...   
2305  26ff4c29        SK Raina       1dc12ab9  ...     0       0         NaN   
2306  26ff4c29        SK Raina       1dc12ab9  ...     0       0         NaN   
2307  26ff4c29        SK Raina       1dc12ab9  ...     0       0         NaN   
2308  26ff4c29      KD Karthik       c03f1114  ...     0       0         NaN   
2309  26ff4c29        SK Raina       1dc12ab9  ...     0       0         NaN   

      player_out_id  fielders_name fielders_id wicket_type runs_batter  \
0               NaN            NaN         NaN         NaN           0   
1               NaN            NaN         NaN         NaN           1   
2               NaN            NaN         NaN         NaN           0   
3               NaN            NaN         NaN         NaN           0   
4               NaN            NaN         NaN         NaN           0   
...             ...            ...         ...         ...         ...   
2305            NaN            NaN         NaN         NaN           6   
2306            NaN            NaN         NaN         NaN           0   
2307            NaN            NaN         NaN         NaN           1   
2308            NaN            NaN         NaN         NaN           1   
2309            NaN            NaN         NaN         NaN           1   

     runs_extras runs_total  
0              0          0  
1              0          1  
2              0          0  
3              0          0  
4              0          0  
...          ...        ...  
2305           0          6  
2306           0          0  
2307           0          1  
2308           0          1  
2309           0          1  

[2310 rows x 23 columns]

## Univariate Analysis

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2310 entries, 0 to 2309
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   game_id         2310 non-null   int64 
 1   season          2310 non-null   object
 2   team            2310 non-null   object
 3   over            2310 non-null   int64 
 4   batter          2310 non-null   object
 5   batter_id       2310 non-null   object
 6   bowler          2310 non-null   object
 7   bowler_id       2310 non-null   object
 8   non_striker     2310 non-null   object
 9   non_striker_id  2310 non-null   object
 10  wides           2310 non-null   int64 
 11  noballs         2310 non-null   int64 
 12  legbyes         2310 non-null   int64 
 13  byes            2310 non-null   int64 
 14  wicket          2310 non-null   int64 
 15  player_out      139 non-null    object
 16  player_out_id   139 non-null    object
 17  fielders_name   101 non-null    object
 18  fielders

In [4]:
data.describe()

game_id         over        wides      noballs      legbyes  \
count    2310.000000  2310.000000  2310.000000  2310.000000  2310.000000   
mean   229012.738528     8.986580     0.032900     0.007359     0.032468   
std     15083.365651     5.692822     0.211718     0.090413     0.234123   
min    211028.000000     0.000000     0.000000     0.000000     0.000000   
25%    211048.000000     4.000000     0.000000     0.000000     0.000000   
50%    225271.000000     9.000000     0.000000     0.000000     0.000000   
75%    238195.000000    14.000000     0.000000     0.000000     0.000000   
max    255954.000000    19.000000     5.000000     2.000000     4.000000   

              byes       wicket  runs_batter  runs_extras   runs_total  
count  2310.000000  2310.000000  2310.000000  2310.000000  2310.000000  
mean      0.006926     0.060173     1.205195     0.079654     1.284848  
std       0.140975     0.237859     1.564374     0.351538     1.548490  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     1.000000     0.000000     1.000000  
75%       0.000000     0.000000     1.000000     0.000000     1.000000  
max       4.000000     1.000000     6.000000     5.000000     7.000000

The dataset contains 22 columns and 2090 observations related to cricket. The ratio of quantitative to categorical (ordinal and nominal) data is approximately 50:50. Note that the columns **player_out, player_out_id, fielder_name, fielders_id, and wicket_type** show a significant amount of missing values. Player_out and player_out_id may be missing due to inapplicability (e.g., no one is out), which results in fielder_name, fielder_id, and wicket_type being invalid in these observations. Thus, these missing data are likely classified as **missing at random(MAR)**. Given the complexity of imputing these values, they are unlikely to be included in our further analysis.

In [5]:
def vis_bar(x_input):
    return alt.Chart(data).mark_bar().encode(
        x = x_input,
        y = "count()"
    ).properties(
        width = 150, 
        height = 150
    )

over = vis_bar("over")
wides = vis_bar("wides")
noballs = vis_bar("noballs")
legbyes = vis_bar("legbyes")
byes = vis_bar("byes")
wicket = vis_bar("wicket")
run_batter = vis_bar("runs_batter")
run_extras = vis_bar("runs_extras")
run_total = vis_bar("runs_total") 

v1 = alt.hconcat(over, wides, noballs)
v2 = alt.hconcat(legbyes, byes, wicket)
v3 = alt.hconcat(run_batter, run_extras, run_total)
alt.vconcat(v1, v2, v3)

alt.VConcatChart(...)

In [6]:
season = vis_bar("season")
team = vis_bar("team")

alt.hconcat(season, team)

alt.HConcatChart(...)

In [7]:
data["batter"].value_counts().reset_index()

batter  count
0    ME Trescothick    173
1          GC Smith    111
2        RT Ponting     88
3         DR Martyn     81
4    PD Collingwood     70
..              ...    ...
96       SI Mahmood      1
97         KD Mills      1
98    CK Langeveldt      1
99      MS Sinclair      1
100       M Zondeki      1

[101 rows x 2 columns]

In [8]:
data["bowler"].value_counts().reset_index()

bowler  count
0              B Lee     85
1          A Symonds     79
2         GD McGrath     74
3            D Gough     62
4      MS Kasprowicz     61
..               ...    ...
68        TM Dilshan     12
69          GC Smith     12
70  DPMD Jayawardene      6
71         JH Kallis      6
72         LPC Silva      6

[73 rows x 2 columns]

In [9]:
data["non_striker"].value_counts().reset_index()

non_striker  count
0   ME Trescothick    199
1         GC Smith    106
2       RT Ponting     86
3        DR Martyn     74
4            B Lee     71
..             ...    ...
95        DR Smith      2
96       DR Tuffey      1
97   CK Langeveldt      1
98    IDR Bradshaw      1
99         J Botha      1

[100 rows x 2 columns]

Please be aware that all columns related to "ID" are not included in this analysis as they are primarily used for identification purposes and are unlikely to exhibit significant patterns. As observed, none of the quantitative attributes follow a normal distribution; the majority of them display right-skewed distributions. An interesting observation is found in the variable "Over" where the frequency decreases with increasing values, warranting further analysis. Regarding ordinal attributes, the dataset contains notably more data from the 2005/06 season compared to other seasons, as well as more teams from "New Zealand," "South Africa," and "Australia" than teams from other continents. Lastly, concerning nominal attributes, GC Smith appears to have considerably more observations than other players in both the "non-striker" and "batter" roles.

In general, the dataset we possess does not form a normal distribution, with samples unevenly distributed across categories. It is important to keep this in mind when training, testing, and interpreting the model, and to appropriately acknowledge this as a potential limitation.

## Multivariate Analysis - Correlation Analysis

In [10]:
corr_data = data[["over", "wides", "noballs", "legbyes", "byes", "wicket", "runs_batter", "runs_extras", "runs_total"]]
corr_df = corr_data.corr()
corr_df

over     wides   noballs   legbyes      byes    wicket  \
over         1.000000 -0.073296 -0.023368 -0.004222 -0.017692  0.063285   
wides       -0.073296  1.000000 -0.012654 -0.021559 -0.007638 -0.039329   
noballs     -0.023368 -0.012654  1.000000 -0.011293 -0.004001 -0.020601   
legbyes     -0.004222 -0.021559 -0.011293  1.000000 -0.006816 -0.035097   
byes        -0.017692 -0.007638 -0.004001 -0.006816  1.000000 -0.012435   
wicket       0.063285 -0.039329 -0.020601 -0.035097 -0.012435  1.000000   
runs_batter  0.085510 -0.119770  0.004629 -0.106883 -0.037868 -0.194979   
runs_extras -0.060060  0.581585  0.240446  0.647375  0.390856 -0.057346   
runs_total   0.072752  0.011033  0.059262  0.038987  0.050476 -0.209998   

             runs_batter  runs_extras  runs_total  
over            0.085510    -0.060060    0.072752  
wides          -0.119770     0.581585    0.011033  
noballs         0.004629     0.240446    0.059262  
legbyes        -0.106883     0.647375    0.038987  
byes           -0.037868     0.390856    0.050476  
wicket         -0.194979    -0.057346   -0.209998  
runs_batter     1.000000    -0.157312    0.974545  
runs_extras    -0.157312     1.000000    0.068094  
runs_total      0.974545     0.068094    1.000000

In [11]:
corr_ = corr_df.stack()
corr_ = corr_.reset_index()
corr_.columns = ['row', 'column', 'corr']
corr_

alt.Chart(corr_).mark_rect().encode(
    x = 'column',
    y = 'row',
    color = 'corr:Q',
    tooltip = 'corr:Q'
).properties(
    width = 400,
    height = 400
)

alt.Chart(...)

In [12]:
corr_table = corr_[corr_["row"] != corr_["column"]]
corr_table.sort_values("corr", ascending = False).head(15)

row       column      corr
78   runs_total  runs_batter  0.974545
62  runs_batter   runs_total  0.974545
34      legbyes  runs_extras  0.647375
66  runs_extras      legbyes  0.647375
16        wides  runs_extras  0.581585
64  runs_extras        wides  0.581585
43         byes  runs_extras  0.390856
67  runs_extras         byes  0.390856
25      noballs  runs_extras  0.240446
65  runs_extras      noballs  0.240446
54  runs_batter         over  0.085510
6          over  runs_batter  0.085510
8          over   runs_total  0.072752
72   runs_total         over  0.072752
79   runs_total  runs_extras  0.068094

In [13]:
def vis_corr_scatter(x_input, y_input):
    return alt.Chart(corr_data).mark_point().encode(
        x = x_input,
        y = y_input
    ).properties(
        width = 150,
        height = 150
    )
    
batter_total_run_total = vis_corr_scatter("runs_batter:Q", "runs_total:Q")
legbyes_runs_extras = vis_corr_scatter("legbyes:Q", "runs_extras:Q")
wides_runs_extras = vis_corr_scatter("wides:Q", "runs_extras:Q")
byes_runs_extras = vis_corr_scatter("byes:Q", "runs_extras:Q")
noballs_runs_extras = vis_corr_scatter("noballs:Q", "runs_extras:Q")
alt.hconcat(batter_total_run_total, legbyes_runs_extras, wides_runs_extras, byes_runs_extras, noballs_runs_extras)

alt.HConcatChart(...)

Among the quantitative attributes, the top 5 relationships in terms of strengths are:
1. **batter_total & run_total(r = 0.97)**
2. **legbyes & runs_extras(r = 0.64)**
3. **wides & runs_extra(r = 0.56**
4. **byes & runs_extra(r = 0.39)**
5. **noballs & run_extra(r = 0.31)**

Note that, using Pearson's r might not be able to capture non-linear relationships that potentially exist among these relationships. 